agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp

```diff
@@ -0,0 +1,111 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include "ap_autoconfiguration_task.h"
+
+#include "../agent_db.h"
+#include "../backhaul_manager/backhaul_manager_thread.h"
+
+#include <tlvf/ieee_1905_1/tlvAlMacAddress.h>
+#include <tlvf/ieee_1905_1/tlvAutoconfigFreqBand.h>
+#include <tlvf/ieee_1905_1/tlvSearchedRole.h>
+#include <tlvf/wfa_map/tlvSearchedService.h>
+#include <tlvf/wfa_map/tlvSupportedService.h>
+
+#include <beerocks/tlvf/beerocks_message_control.h>
+
+#include <easylogging++.h>
+
+using namespace beerocks;
+using namespace net;
+using namespace son;
+
+static constexpr uint8_t MAX_FAILED_AUTOCONFIG_SEARCH_ATTEMPTS = 20;
+static constexpr uint8_t AUTOCONFIG_DISCOVERY_TIMEOUT_SECONDS  = 1;
+
+#define FSM_MOVE_STATE(radio_iface, new_state)                                                     \
+    ({                                                                                             \
+        LOG(TRACE) << "AP_AUTOCONFIGURATION " << radio_iface                                       \
+                   << " FSM: " << fsm_state_to_string(m_state[radio_iface].state) << " --> "       \
+                   << fsm_state_to_string(new_state);                                              \
+        m_state[radio_iface].state = new_state;                                                    \
+    })
+
+const std::string ApAutoConfigurationTask::fsm_state_to_string(eState status)
+{
+    switch (status) {
+    case eState::UNCONFIGURED:
+        return "UNCONFIGURED";
+    case eState::CONTROLLER_DISCOVERY:
+        return "CONTROLLER_DISCOVERY";
+    case eState::WAIT_FOR_CONTROLLER_DISCOVERY_COMPLETE:
+        return "WAIT_FOR_CONTROLLER_DISCOVERY_COMPLETE";
+    case eState::AP_CONFIGURATION:
+        return "AP_CONFIGURATION";
+    case eState::WAIT_FOR_AP_CONFIGURATION_COMPLETE:
+        return "WAIT_FOR_AP_CONFIGURATION_COMPLETE";
+    case eState::CONFIGIRED:
+        return "CONFIGIRED";
+    default:
+        LOG(ERROR) << "state argument doesn't have an enum";
+        break;
+    }
+    return std::string();
+}
```

Comment:
It backhaul manager string representation is generated by `const char *backhaul_manager::s_arrStates[] = {FOREACH_STATE(GENERATE_STRING)};`.
Let's unify the code :)

Reply:
TBH I hate the whole "FOREACH_STATE(GENERATE_STRING)}" macro on the backhaul manager. We suppose to write code in CPP and not in MACRO.
Using macro should take place only if it has a good purpose.
The only reason I have chosen to use "FSM_MOVE_STATE" is because I wanted that on a state switch, it will print to the log file the line that the switch has occurred. If it was a regular function it would print the same line everywhere.

I'll might unify the code by removing this macro crap on the backhaul manager as part of the unified agent epic. 🙂